File: solutions/binary-tree-pruning/binary-tree-pruning.go
// @Title: Binary Tree Pruning
// @Author: colinjxc
// @Date: 2022-03-01T21:31:00+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-pruning


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pruneTree(root *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }
    root.Left = pruneTree(root.Left)
    root.Right = pruneTree(root.Right)
    if root.Left == nil && root.Right == nil && root.Val == 0 {
        return nil
    }
    return root;
}

File: solutions/binary-tree-pruning/binary-tree-pruning.cpp
// @Title: Binary Tree Pruning
// @Author: colinjxc
// @Date: 2022-03-01T21:26:56+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-pruning


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        if (root == nullptr) {
            return nullptr;
        }
        root->right = pruneTree(root->right);
        root->left = pruneTree(root->left);
        if (!root->right && !root->left && !root->val) {
            return nullptr;
        }
        return root;
    }
};
